File: src/data/types/paymentResult.ts
import type { CrewMember } from "./crewMember.ts";
import type { Wallet } from "./wallet.ts";

// Necessary completion: Agent 16's contract names `payUpkeep(...):
// PaymentResult` but never defines it. Three outcomes, not two -- "not
// due yet" (called before a full WAGE_PAYMENT_INTERVAL_HOURS has
// elapsed) is a normal no-op, distinct from an actual failure
// (insufficient funds), which is what feeds checkAttrition's grace-period
// clock via an unmoved lastPaidAt.
export interface PaymentPaid {
  status: "paid";
  updatedCrewMember: CrewMember;
  updatedWallet: Wallet;
}

export interface PaymentNotDue {
  status: "not-due";
}

export interface PaymentInsufficientFunds {
  status: "insufficient-funds";
}

export type PaymentResult = PaymentPaid | PaymentNotDue | PaymentInsufficientFunds;

File: unity/ProfitableCore/Schema/PaymentResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/paymentResult.ts's PaymentResult = PaymentPaid |
// PaymentNotDue | PaymentInsufficientFunds discriminated union (three
// outcomes, not two -- "not due yet" is a normal no-op, distinct from an
// actual failure).
public abstract class PaymentResult
{
    public abstract string Status { get; }
}

public sealed class PaymentPaid : PaymentResult
{
    public override string Status => "paid";
    public CrewMember UpdatedCrewMember { get; init; } = new();
    public Wallet UpdatedWallet { get; init; } = new();
}

public sealed class PaymentNotDue : PaymentResult
{
    public override string Status => "not-due";
}

public sealed class PaymentInsufficientFunds : PaymentResult
{
    public override string Status => "insufficient-funds";
}
